src/JSON.jl

```diff
@@ -10,6 +10,47 @@ include("Parser.jl")
 
 import .Parser.parse
 
+typealias JSONPrimitive @compat(Union{
+        Associative, Tuple, AbstractArray, AbstractString, Integer,
+        AbstractFloat, Bool, Void})
+
+"""
+Return a value of a JSON-encodable primitive type that `x` should be lowered
+into before encoding as JSON. Supported types are: `Associative` to JSON
+objects, `Tuple` and `AbstractVector` to JSON arrays, `AbstractArray` to nested
+JSON arrays, `AbstractString` to JSON string, `Integer` and `AbstractFloat` to
+JSON number, `Bool` to JSON boolean, and `Void` to JSON null.
+
+Extensions of this method should preserve the property that the return value is
+one of the aforementioned types. If first lowering to some intermediate type is
+required, then extensions should call `lower` before returning a value.
+
+Note that the return value need not be *recursively* lowered—this function may
+for instance return an `AbstractArray{Any, 1}` whose elements are not JSON
+primitives.
+"""
+function lower(a)
+    obj = @compat Dict{Symbol, Any}()
+    for name in @compat fieldnames(a)
+        obj[name] = getfield(a, name)
+    end
+    obj
+end
```

Comment:
This definition seems quite broad as a fallback.  

For example, if I [define a new bitstype](https://github.com/diegozea/BioSeq.jl/blob/master/src/biounits.jl#L1-L4), it will fallback here, which is probably not that useful.  Maybe this can be restricted to `DataTypes`?  (Not sure...)